mk3/launchpad.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2013 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mk3 provides interfaces to talk to Novation Launchpad MK3 via MIDI in and out.
+// Tested with a Launchpad Mini MK3
+package mk3
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/rakyll/portmidi"
+)
+
+// Launchpad represents a device with an input and output MIDI stream.
+type Launchpad struct {
+	inputStream  *portmidi.Stream
+	outputStream *portmidi.Stream
+	Block        bool
+}
+
+// Hit represents physical touches to Launchpad buttons.
+type Hit struct {
+	X int
+	Y int
+}
+
+// Open opens a connection Launchpad and initializes an input and output
+// stream to the currently connected device. If there are no
+// devices are connected, it returns an error.
+func Open() (*Launchpad, error) {
+	input, output, err := discover()
+	if err != nil {
+		return nil, err
+	}
+
+	var inStream, outStream *portmidi.Stream
+	if inStream, err = portmidi.NewInputStream(input, 1024); err != nil {
+		return nil, err
+	}
+	if outStream, err = portmidi.NewOutputStream(output, 1024, 0); err != nil {
+		return nil, err
+	}
+	// Switch to the session mode.
+	outStream.WriteSysExBytes(portmidi.Time(), []byte{0xf0, 0x00, 0x20, 0x29, 0x02, 0x18, 0x22, 0x00, 0xf7})
+	return &Launchpad{inputStream: inStream, outputStream: outStream}, nil
+}
+
+// Listen listens the input stream for hits.
+func (l *Launchpad) Listen() <-chan Hit {
+	ch := make(chan Hit)
+	go func(pad *Launchpad, ch chan Hit) {
+		for {
+
+			if !l.Block {
+				// sleep for a while before the new polling tick,
+				// otherwise operation is too intensive and blocking
+				time.Sleep(10 * time.Millisecond)
+				hits, err := pad.Read()
+				if err != nil {
+					continue
+				}
+				for i := range hits {
+					ch <- hits[i]
+				}
+			} else {
+				pad.Read() // Throw away whilst we're blocked.
+			}
+		}
+	}(l, ch)
+	return ch
+}
+
+func (l *Launchpad) BlockKeys(block bool) {
```

Comment:
Do we expect this to be safe concurrent use?

Reply:
removed the block keys feature as this was not my problem I was trying to solve.